File: android/src/main/java/com/framepayments/reactnativeframe/ProveAuthBridge.kt
package com.framepayments.reactnativeframe

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableNativeMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.prove.sdk.proveauth.AuthFinishStep
import com.prove.sdk.proveauth.OtpFinishInput
import com.prove.sdk.proveauth.OtpFinishStep
import com.prove.sdk.proveauth.OtpStartInput
import com.prove.sdk.proveauth.OtpStartStep
import com.prove.sdk.proveauth.ProveAuth
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

// Android bridge for Prove's ProveAuth SDK. Surface mirrors the iOS bridge:
//
//   authenticate(authToken) -> Promise<{ status: 'success' | 'failed' }>
//   submitOtp(code) -> void   (resumes a suspended OTP request)
//   cancelOtp() -> void
//   cancelAuth() -> void
//
//   Event: FrameProveOtpNeeded — fires when Prove falls back to OTP.
//
// The Prove Android SDK (com.prove.sdk:proveauth) is declared as compileOnly
// in our android/build.gradle so this file compiles without forcing host apps
// that don't ship phone_verification to pull it in. Host apps that DO ship
// phone_verification add `implementation 'com.prove.sdk:proveauth:<version>'`
// to their own app/build.gradle. At runtime, we probe for the SDK class and
// resolve PROVE_UNAVAILABLE if missing so JS falls back to the Frame OTP path.
//
// Call sequence verified against frame-android's ProveAuthService.kt.

class ProveAuthBridge(private val reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  private val scope = CoroutineScope(Dispatchers.Default + Job())
  private var authPromise: Promise? = null
  private var otpDeferred: CompletableDeferred<String?>? = null

  // Retained across the auth lifecycle so the Prove SDK's step closures don't
  // outlive the bridge's reference to them; cleared on terminal resolution.
  private var proveAuth: ProveAuth? = null

  override fun getName(): String = "FrameProveAuth"

  // RTNDeviceEventEmitter is the documented mechanism for arbitrary RN events
  // from a non-RCTEventEmitter module. JS subscribes via NativeEventEmitter.
  @ReactMethod
  fun addListener(@Suppress("UNUSED_PARAMETER") eventName: String) {}

  @ReactMethod
  fun removeListeners(@Suppress("UNUSED_PARAMETER") count: Int) {}

  // ─── JS API ───

  @ReactMethod
  fun authenticate(authToken: String, promise: Promise) {
    if (authPromise != null) {
      promise.reject("PROVE_BUSY", "A Prove authentication is already in flight.")
      return
    }

    if (!isProveSdkLinked()) {
      promise.reject("PROVE_UNAVAILABLE", "Prove SDK is not linked in this app.")
      return
    }

    authPromise = promise
    scope.launch {
      try {
        runProveFlow(authToken)
        finishAuthSuccess()
      } catch (e: NoClassDefFoundError) {
        finishAuthFailure("Prove SDK is not linked in this app.")
      } catch (e: Throwable) {
        finishAuthFailure(e.message ?: "Prove authentication failed.")
      }
    }
  }

  @ReactMethod
  fun submitOtp(code: String, promise: Promise) {
    val deferred = otpDeferred
    if (deferred == null) {
      promise.resolve(null)
      return
    }
    otpDeferred = null
    deferred.complete(code)
    promise.resolve(null)
  }

  @ReactMethod
  fun cancelOtp(promise: Promise) {
    otpDeferred?.complete(null)
    otpDeferred = null
    promise.resolve(null)
  }

  @ReactMethod
  fun cancelAuth(promise: Promise) {
    otpDeferred?.complete(null)
    otpDeferred = null
    finishAuthFailure("Prove authentication canceled by user.")
    promise.resolve(null)
  }

  override fun invalidate() {
    scope.cancel()
    super.invalidate()
  }

  // ─── Internal ───

  private fun isProveSdkLinked(): Boolean = try {
    Class.forName("com.prove.sdk.proveauth.ProveAuth")
    true
  } catch (_: Throwable) {
    false
  }

  private fun emitOtpNeeded() {
    val params = WritableNativeMap()
    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("FrameProveOtpNeeded", params)
  }

  private suspend fun runProveFlow(authToken: String) {
    val resultDeferred = CompletableDeferred<Result<Boolean>>()

    // AuthFinishStep fires when Prove signals "device auth ok". JS owns the
    // backend confirm step, so we resolve the bridge promise as soon as the
    // SDK reaches this point.
    val authFinishStep = AuthFinishStep { _ ->
      resultDeferred.complete(Result.success(true))
    }

    // The phone number is encoded in the authToken; phoneNumberNeeded=true is
    // a misconfiguration → error. Mirrors frame-android ProveAuthService.kt.
    val otpStartStep = OtpStartStep { phoneNumberNeeded, _, callback ->
      if (phoneNumberNeeded) {
        callback.onError()
      } else {
        callback.onSuccess(OtpStartInput(""))
      }
    }

    // OtpFinishStep — Prove falls back to OTP. Emit the event so JS shows its
    // entry sheet, then suspend on otpDeferred until submitOtp/cancelOtp
    // resolves it. Null means the user cancelled.
    val otpFinishStep = OtpFinishStep { _, callback ->
      scope.launch {
        // Guard against a stale deferred from a previous OTP request (Prove
        // can re-request OTP if the first code fails validation).
        otpDeferred?.complete(null)
        val deferred = CompletableDeferred<String?>()
        otpDeferred = deferred
        emitOtpNeeded()
        val otp = deferred.await()
        otpDeferred = null
        if (otp != null) {
          callback.onSuccess(OtpFinishInput(otp))
        } else {
          callback.onError()
        }
      }
    }

    val sdk = ProveAuth.builder()
      .withContext(reactContext.applicationContext)
      .withAuthFinishStep(authFinishStep)
      .withOtpFallback(otpStartStep, otpFinishStep)
      .build()
    proveAuth = sdk

    try {
      sdk.authenticate(authToken)
    } catch (e: Throwable) {
      releaseRetainedSDKObjects()
      resultDeferred.complete(Result.failure(e))
    }

    resultDeferred.await().getOrThrow()
    releaseRetainedSDKObjects()
  }

  private fun releaseRetainedSDKObjects() {
    proveAuth = null
  }

  private fun finishAuthSuccess() {
    val promise = authPromise ?: return
    authPromise = null
    val result = WritableNativeMap().apply { putString("status", "success") }
    promise.resolve(result)
  }

  private fun finishAuthFailure(message: String) {
    val promise = authPromise ?: return
    authPromise = null
    val result = WritableNativeMap().apply {
      putString("status", "failed")
      putString("message", message)
    }
    promise.resolve(result)
  }
}

File: ios/ProveAuthBridge.swift
//
//  ProveAuthBridge.swift
//  FrameReactNative
//
//  Bridges Prove's iOS ProveAuth SDK to React Native. The JS surface:
//
//   - authenticate(authToken: String) → Promise<{ status: 'success' | 'failed' }>
//   - submitOtp(code: String) → void   (call when JS shows the OTP fallback UI)
//   - cancelOtp() → void
//   - cancelAuth() → void              (force-cancel the in-flight authentication)
//
//   Event: `FrameProveOtpNeeded` — fires when the Prove SDK falls back to OTP.
//   JS listens for the event, shows the OTP entry sheet, and calls submitOtp.
//
//  The Frame backend's confirmVerification call is NOT performed here — JS
//  handles it via PhoneVerificationsAPI after authenticate() resolves success.
//  We pass a no-op confirmHandler to AuthFinishStep so the SDK's state machine
//  resolves the authenticate promise as soon as Prove finishes its part.
//
//  ProveAuth is loaded behind `#if canImport(ProveAuth)`; when the SDK isn't
//  linked, authenticate() resolves PROVE_UNAVAILABLE so JS can fall back to
//  the Frame OTP path.
//
//  Protocol shapes are verified against the Frame iOS 3.0.0 reference at
//  Sources/FrameOnboarding/Services/ProveAuthService.swift:
//    - ProveAuthFinishStep.execute(authId:)
//    - OtpStartStep.execute(phoneNumberNeeded:phoneValidationError:callback:)
//    - OtpFinishStep.execute(otpError:callback:)
//

import Foundation
import React

#if canImport(ProveAuth)
import ProveAuth
#endif

@objc(FrameProveAuth)
public class FrameProveAuth: RCTEventEmitter {

  private static let EVENT_OTP_NEEDED = "FrameProveOtpNeeded"

#if canImport(ProveAuth)
  private var proveAuth: ProveAuth?
  private var authFinishStep: BridgeAuthFinishStep?
  private var otpStartStep: BridgeOtpStartStep?
  private var otpFinishStep: BridgeOtpFinishStep?
#endif

  // Single-resume guard around the JS promise. Both the SDK completion
  // callback and the AuthFinishStep onResult closure can fire — first one
  // wins, the other no-ops via this nil-check.
  private var authResolve: RCTPromiseResolveBlock?
  private var authReject: RCTPromiseRejectBlock?

  // Continuation handed to the OTP-finish step. Captured when the SDK calls
  // OtpFinishStep.execute(...) and resumed by submitOtp/cancelOtp from JS.
  // The Prove SDK can re-request OTP if the first code fails validation, so
  // the bridge guards against an existing continuation by error-resuming the
  // old one before installing the new one.
  private var otpContinuation: ((String?) -> Void)?

  // MARK: - RCTEventEmitter

  public override static func requiresMainQueueSetup() -> Bool {
    return false
  }

  public override func supportedEvents() -> [String] {
    return [FrameProveAuth.EVENT_OTP_NEEDED]
  }

  // MARK: - JS API

  @objc public func authenticate(_ authToken: String,
                                  resolver resolve: @escaping RCTPromiseResolveBlock,
                                  rejecter reject: @escaping RCTPromiseRejectBlock) {
#if canImport(ProveAuth)
    DispatchQueue.main.async { [weak self] in
      guard let self = self else {
        reject("PROVE_BRIDGE_GONE", "Prove bridge was deallocated.", nil)
        return
      }
      guard self.authResolve == nil else {
        reject("PROVE_BUSY", "A Prove authentication is already in flight.", nil)
        return
      }
      self.authResolve = resolve
      self.authReject = reject
      self.startProveAuth(authToken: authToken)
    }
#else
    // Prove SDK not linked in this build. Return a clean failure so JS can
    // fall back to the Frame OTP path without surfacing a crash.
    reject("PROVE_UNAVAILABLE", "Prove SDK is not linked in this app.", nil)
#endif
  }

  @objc public func submitOtp(_ code: String,
                               resolver resolve: @escaping RCTPromiseResolveBlock,
                               rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async { [weak self] in
      guard let self = self else {
        reject("PROVE_BRIDGE_GONE", "Prove bridge was deallocated.", nil)
        return
      }
      guard let cont = self.otpContinuation else {
        // OTP wasn't requested; treat as a no-op so JS error handlers don't
        // double-throw on a stale OTP submit.
        resolve(nil)
        return
      }
      self.otpContinuation = nil
      cont(code)
      resolve(nil)
    }
  }

  @objc public func cancelOtp(_ resolve: @escaping RCTPromiseResolveBlock,
                               rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async { [weak self] in
      self?.otpContinuation?(nil)
      self?.otpContinuation = nil
      resolve(nil)
    }
  }

  @objc public func cancelAuth(_ resolve: @escaping RCTPromiseResolveBlock,
                                rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async { [weak self] in
      self?.otpContinuation?(nil)
      self?.otpContinuation = nil
      self?.handleAuthFailure(message: "Prove authentication canceled by user.")
      resolve(nil)
    }
  }

  // MARK: - Prove orchestration

#if canImport(ProveAuth)
  private func startProveAuth(authToken: String) {
    // AuthFinishStep fires when Prove signals "device auth ok"; JS owns the
    // backend confirm step, so we hand the SDK a no-op confirmHandler that
    // returns success immediately.
    let bridge = BridgeWeakBox(self)

    self.authFinishStep = BridgeAuthFinishStep(
      confirmHandler: { _, _ in /* JS owns the confirm step */ },
      onResult: { result in
        DispatchQueue.main.async {
          bridge.value?.handleAuthResult(result)
        }
      }
    )
    self.otpStartStep = BridgeOtpStartStep(onOtpNeeded: {
      DispatchQueue.main.async {
        bridge.value?.emitOtpNeeded()
      }
    })
    self.otpFinishStep = BridgeOtpFinishStep(otpProvider: {
      // Suspend until submitOtp/cancelOtp resolves the continuation. If a
      // prior continuation is somehow still live (Prove re-requested OTP
      // before the first one was answered), fail the old one so it doesn't
      // hang the SDK state machine.
      return await withCheckedContinuation { cont in
        DispatchQueue.main.async {
          if let stale = bridge.value?.otpContinuation {
            bridge.value?.otpContinuation = nil
            stale(nil)
          }
          bridge.value?.otpContinuation = { code in cont.resume(returning: code) }
        }
      }
    })

    guard let authFinishStep = self.authFinishStep,
          let otpStartStep = self.otpStartStep,
          let otpFinishStep = self.otpFinishStep else { return }

    self.proveAuth = ProveAuth.builder(authFinish: authFinishStep)
      .withOtpFallback(otpStart: otpStartStep, otpFinish: otpFinishStep)
      .build()

    DispatchQueue.global(qos: .userInitiated).async { [weak self] in
      guard let prove = self?.proveAuth else { return }
      prove.authenticate(authToken: authToken) { error in
        // Some Prove SDK versions deliver completion via AuthFinishStep;
        // others surface terminal errors here. Both paths converge on the
        // single-resume handleAuthFailure / handleAuthResult guard.
        if error.errorDescription != nil {
          DispatchQueue.main.async {
            self?.handleAuthFailure(message: error.errorDescription ?? "Prove authentication failed.")
          }
        }
      }
    }
  }

  private func handleAuthResult(_ result: Result<Bool, Error>) {
    switch result {
    case .success:
      finishAuthSuccess()
    case .failure(let error):
      handleAuthFailure(message: error.localizedDescription)
    }
  }
#endif

  private func emitOtpNeeded() {
    self.sendEvent(withName: FrameProveAuth.EVENT_OTP_NEEDED, body: [:])
  }

  private func finishAuthSuccess() {
    guard let resolve = self.authResolve else { return }
    self.authResolve = nil
    self.authReject = nil
    resolve(["status": "success"])
    // Defer SDK release until after the JS promise has resolved so any
    // in-flight step closures aren't reading dangling references.
    DispatchQueue.main.async { [weak self] in
      self?.releaseRetainedSDKObjects()
    }
  }

  private func handleAuthFailure(message: String) {
    guard let resolve = self.authResolve else { return }
    self.authResolve = nil
    self.authReject = nil
    resolve(["status": "failed", "message": message])
    DispatchQueue.main.async { [weak self] in
      self?.releaseRetainedSDKObjects()
    }
  }

  private func releaseRetainedSDKObjects() {
#if canImport(ProveAuth)
    self.proveAuth = nil
    self.authFinishStep = nil
    self.otpStartStep = nil
    self.otpFinishStep = nil
#endif
    self.otpContinuation = nil
  }
}

// Weak wrapper used to thread the bridge through Prove SDK step closures
// without creating retain cycles.
private final class BridgeWeakBox<T: AnyObject> {
  weak var value: T?
  init(_ value: T) { self.value = value }
}

#if canImport(ProveAuth)

// MARK: - Prove SDK step adapters (mirror Frame iOS ProveAuthService.swift)

/// AuthFinishStep — fires `onResult(.success)` after Prove signals the device
/// auth is complete and our (no-op) confirmHandler returns.
private final class BridgeAuthFinishStep: NSObject, ProveAuthFinishStep, @unchecked Sendable {
  private let confirmHandler: @Sendable (String, String) async throws -> Void
  private let onResult: @Sendable (Result<Bool, Error>) -> Void

  init(
    confirmHandler: @escaping @Sendable (String, String) async throws -> Void,
    onResult: @escaping @Sendable (Result<Bool, Error>) -> Void
  ) {
    self.confirmHandler = confirmHandler
    self.onResult = onResult
  }

  func execute(authId: String) {
    Task {
      do {
        try await confirmHandler("", "")
        onResult(.success(true))
      } catch {
        onResult(.failure(error))
      }
    }
  }
}

/// OtpStartStep — Prove tells us OTP fallback is starting. The phone number
/// is already known to Prove (it came from authToken's claims), so we accept
/// the start; phoneNumberNeeded=true is a misconfiguration → error.
private final class BridgeOtpStartStep: NSObject, OtpStartStep, @unchecked Sendable {
  private let onOtpNeeded: @Sendable () -> Void
  init(onOtpNeeded: @escaping @Sendable () -> Void) { self.onOtpNeeded = onOtpNeeded }

  func execute(phoneNumberNeeded: Bool, phoneValidationError: ProveAuthError?, callback: any OtpStartStepCallback) {
    if phoneNumberNeeded {
      callback.onError()
    } else {
      onOtpNeeded()
      callback.onSuccess(input: nil)
    }
  }
}

/// OtpFinishStep — Prove asks the host to collect the OTP from the user. We
/// await JS via the continuation provider; null means cancel.
private final class BridgeOtpFinishStep: NSObject, OtpFinishStep, @unchecked Sendable {
  private let otpProvider: @Sendable () async -> String?
  init(otpProvider: @escaping @Sendable () async -> String?) { self.otpProvider = otpProvider }

  func execute(otpError: ProveAuthError?, callback: any OtpFinishStepCallback) {
    Task { @MainActor in
      let otp = await otpProvider()
      if let otp = otp {
        callback.onSuccess(input: OtpFinishInput(otp: otp))
      } else {
        callback.onError()
      }
    }
  }
}

#endif
